x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/AutoFollowCoordinator.java

```diff
@@ -0,0 +1,293 @@
+                    public void onFailure(String source, Exception e) {
+                        handler.accept(e);
+                    }
+
+                    @Override
+                    public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+                        handler.accept(null);
+                    }
+                });
+            }
+
+        };
+        operation.autoFollowIndices();
+    }
+
+    @Override
+    public void applyClusterState(ClusterChangedEvent event) {
+        final boolean beforeLocalMasterNode = localNodeMaster;
+        localNodeMaster = event.localNodeMaster();
+        if (beforeLocalMasterNode == false && localNodeMaster) {
+            threadPool.schedule(pollInterval, ThreadPool.Names.SAME, this::doAutoFollow);
+        }
+    }
+
+    abstract static class AutoFollower {
+
+        private final Client client;
+        private final Consumer<Exception> handler;
+        private final AutoFollowMetadata autoFollowMetadata;
+
+        private final CountDown countDown;
+        private final AtomicReference<Exception> errorHolder = new AtomicReference<>();
+
+        AutoFollower(Client client, Consumer<Exception> handler, AutoFollowMetadata autoFollowMetadata) {
+            this.client = client;
+            this.handler = handler;
+            this.autoFollowMetadata = autoFollowMetadata;
+            this.countDown = new CountDown(autoFollowMetadata.getPatterns().size());
+        }
+
+        void autoFollowIndices() {
+            for (Map.Entry<String, AutoFollowPattern> entry : autoFollowMetadata.getPatterns().entrySet()) {
+                String clusterAlias = entry.getKey();
+                AutoFollowPattern autoFollowPattern = entry.getValue();
+                Client leaderClient = clusterAlias.equals("_local_") ? client : client.getRemoteClusterClient(clusterAlias);
+                List<String> followedIndices = autoFollowMetadata.getFollowedLeaderIndexUUIDs().get(clusterAlias);
+
+                getLeaderClusterState(leaderClient, (leaderClusterState, e) -> {
+                    if (leaderClusterState != null) {
+                        assert e == null;
+                        handleClusterAlias(clusterAlias, autoFollowPattern, followedIndices, leaderClusterState);
+                    } else {
+                        finalise(e);
+                    }
+                });
+            }
+        }
+
+        private void handleClusterAlias(String clusterAlias, AutoFollowPattern autoFollowPattern,
+                                List<String> followedIndexUUIDs, ClusterState leaderClusterState) {
+            List<Index> leaderIndicesToFollow = getLeaderIndicesToFollow(autoFollowPattern, leaderClusterState, followedIndexUUIDs);
+            if (leaderIndicesToFollow.isEmpty()) {
+                finalise(null);
+            } else {
+                CountDown countDown = new CountDown(leaderIndicesToFollow.size());
```

Comment:
Reusing the name `countDown` scares me here because it shadows the `countDown` from the `AutoFollower`. It is especially confusing with this block of code:

```java
                            if (countDown.countDown()) {
                                finalise(updateError);
                            }
```

where that is counting down the shadower (the local `countDown` defined in `handleClusterAlias` and then `finalise` itself counts down another `CountDown` named `countDown` (the `AutoFollower` `CountDown`). Can we give these clearer names, and definitely not shadow?

Reply:
Agreed. I'll rename them.